packages/@aws-cdk/aws-events/lib/connection.ts

```diff
@@ -0,0 +1,449 @@
+}
+
+/**
+ * Contains the API key authorization parameters to use for the connection.
+ */
+export interface ApiKeyAuthParameters {
+  /**
+   * The name of the API key to use for authorization.
+   */
+  readonly apiKeyName: string;
+
+  /**
+   * The value for the API key to use for authorization.
+   */
+  readonly apiKeyValue: string;
+}
+
+/**
+ * Contains the Basic authorization parameters to use for the connection.
+ */
+export interface BasicAuthParameters {
+  /**
+   * The password associated with the user name to use for Basic authorization.
+   */
+  readonly password: string;
+
+  /**
+   * The user name to use for Basic authorization.
+   */
+  readonly username: string;
+}
+
+/**
+ * Contains the client parameters for OAuth authorization.
+ */
+export interface ClientParameters {
+  /**
+   * The client secret associated with the client ID to use for OAuth authorization for the connection.
+   */
+  readonly clientID: string;
+
+  /**
+   * The client ID to use for OAuth authorization for the connection.
+   */
+  readonly clientSecret: string;
+}
+
+/**
+ * Additional parameters to include with the connection.
+ */
+export interface Parameter {
+  /**
+   * Specifies whether the value is a secret.
+   * @default - True
+   */
+  readonly isValueSecret?: string;
+
+  /**
+   * The key for the parameter.
+   */
+  readonly key: string;
+
+  /**
+   * The value associated with the key for the parameter.
+   */
+  readonly value: string;
+}
+
+/**
+ * These are custom parameter to be used when the target is an API Gateway REST APIs or EventBridge ApiDestinations.
+ */
+export interface HttpParameters {
+  /**
+   * Not currently supported by AWS CloudFormation.
+   */
+  // readonly bodyParameters?: Parameter[];
+
+  /**
+   * The headers that need to be sent as part of request invoking the API Gateway REST API or EventBridge ApiDestination.
+   * @default - none
+   */
+  readonly headerParameters?: Parameter[];
+
+  /**
+   * The query string keys/values that need to be sent as part of request invoking the API Gateway REST API or EventBridge ApiDestination.
+   * @default - none
+   */
+  readonly queryStringParameters?: Parameter[];
+}
+
+/**
+ * Contains the OAuth authorization parameters to use for the connection.
+ */
+export interface OAuthParameters {
+  /**
+   * The URL to the authorization endpoint when OAuth is specified as the authorization type.
+   */
+  readonly authorizationEndpoint: string;
+
+  /**
+   * Contains the client parameters for OAuth authorization.
+   */
+  readonly clientParameters: ClientParameters;
+
+  /**
+   * The method to use for the authorization request.
+   */
+  readonly httpMethod: string;
+
+  /**
+   * Contains the OAuth authorization parameters to use for the connection.
+   * @default - none
+   */
+  readonly oAuthHttpParameters?: HttpParameters;
+}
+
+/**
+ * Contains the authorization parameters to use to authorize with the endpoint.
+ */
+export interface AuthParameters {
```

Comment:
How is this structure supposed to be used? I assume at most one of the auth types needs to be supplied?

In this case, I think I'd prefer you model this as an [enum-like class](https://github.com/aws/aws-cdk/blob/master/docs/DESIGN_GUIDELINES.md#enums):

```ts
export class AuthParameters {
  public static basicAuth(username: string, password: SecretValue): AuthParameters {
    return new AuthParameters({
      authorizationType: AuthorizationType.BASIC,
      basicAuthParameters: { username, password },
    });
  }

  public static ...() {
  }

  constructor(public readonly authParametersJson: any) {
  }
}
```

Reply:
Thanks for the feedback @rix0rrr 
I'm a little confused on how to proceed, is there another similar example in the code base of a enum-like class?
I if use `authParametersJson` as `any` then I end up needing to parse out the `authorisationType` when it comes to building the object
```
let connection = new CfnConnection(this, 'Connection', {
authorizationType: props.authParameters.authParametersJson.authorizationType,
authParameters: props.authParameters.authParametersJson,
description: props.description,
name: this.physicalName,
});
```
I also need to figure out how to build the header and query strings for each connection